src/main/java/com/cloudbees/hudson/plugins/folder/ChildNameGenerator.java

```diff
@@ -0,0 +1,290 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.hudson.plugins.folder;
+
+import com.cloudbees.hudson.plugins.folder.computed.ComputedFolder;
+import hudson.Util;
+import hudson.model.AbstractItem;
+import hudson.model.Action;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.JobProperty;
+import hudson.model.TopLevelItem;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.WeakHashMap;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.model.TransientActionFactory;
+
+/**
+ * Provides a way for a {@link ComputedFolder} may need to break the association between the directory names on disk
+ * that are used to store its items and the {@link Item#getName()} which is used to create the URL of the item.
+ * <p>
+ * <strong>NOTE:</strong> if you need to implement this functionality, you need to ensure that users cannot rename
+ * items within the {@link ComputedFolder} as renaming is not supported when using a {@link ChildNameGenerator}.
+ * <p>
+ * Challenges:
+ * <ul>
+ * <li>See the notes on {@link #itemNameFromItem(AbstractFolder, TopLevelItem)} and
+ * {@link #dirNameFromItem(AbstractFolder, TopLevelItem)} regarding the constraints on how to name things</li>
+ * <li>There are some items which need the {@link Item#getRootDir()} during construction (those are bold evil item types
+ * that leak side-effects, you should fix them if you find them). While you wait for them to be fixed you will need
+ * to work-arround the issue by ensuring that you call {@link #beforeCreateItem(AbstractFolder, String, String)}
+ * passing the {@link Item#getName()} you want the item to have <strong>and</strong> the ideal unmangled name
+ * <strong>before</strong> you call {@code new ChildItemType(parent,name)} and then call
+ * {@link #afterItemCreated(Trace)} when the constructor has returned. Then insure that your
+ * {@link #itemNameFromItem(AbstractFolder, TopLevelItem)} and {@link #dirNameFromItem(AbstractFolder, TopLevelItem)}
+ * fall back to {@link #idealNameFromItem(AbstractFolder, TopLevelItem)} when the magic property they are looking
+ * for is missing.</li>
+ * </ul>
+ *
+ * @param <P> the type of {@link AbstractFolder}.
+ * @param <I> the type of {@link TopLevelItem} within the folder.
+ * @since 5.17
+ */
+// TODO migrate this functionality (by changing the base class) into core once baseline Jenkins has JENKINS-41222 merged
+public abstract class ChildNameGenerator<P extends AbstractFolder<I>, I extends TopLevelItem> {
+    /**
+     * The name of the file that contains the actual name of the child item.
+     */
+    public static final String CHILD_NAME_FILE = "name-utf8.txt";
```

Comment:
I think just `name.txt` would suffice. There is no reason to make the filename so chatty.